### DISCUSSION TITLE
`MikroORM config file not found` after update to `6.3.0`

### DISCUSSION THREAD

**Discussion by sz3lbi** [mikro-orm/mikro-orm#5852] [General]
### Describe the bug

Hello.  
I've updated MikroORM to `6.3.0`. When I try to run the application, I get the following error: `Error: MikroORM config file not found in ['./src/mikro-orm.config.js', './mikro-orm.config.js']`

I placed my config file in `./src/mikro-orm.config.ts` (not `.js`), as described on [this page](https://mikro-orm.io/docs/guide/first-entity), which I followed when setting up my project.
After changing the extension to `.js` and adding `//@ts-check` at the top of the file, it works fine. I guess this can be a breaking change in a minor release?

### Reproduction

Follow the "Getting Started" tutorial. Source of my project that does not start: https://codeberg.org/szelbi/classifieds-service-backend-typescript/src/commit/c057653b8724335d959c7523a6c598452e13bc69

### What driver are you using?

@mikro-orm/postgresql

### MikroORM version

6.3.0

### Node.js version

22.4.1

### Operating system

Debian 12

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/mikro-orm/mikro-orm/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://mikro-orm.io/docs).
- [X] Check that there isn't [already an issue](https://github.com/mikro-orm/mikro-orm/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/mikro-orm/mikro-orm/discussions) or join our [Discord](https://discord.gg/w8bjxFHS7X).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by B4nan** [comment]
There is a working repository with the project which is upgraded to v6.3, try comparing what you do differently.

https://github.com/mikro-orm/guide

> Error: MikroORM config file not found in ['./src/mikro-orm.config.js', './mikro-orm.config.js']

This sounds like `ts-node` is not properly installed (since that is now the only reason TS files would be ignored here), maybe its something related to pnpm? For some reason I cant really use it myself, it never worked for me, always getting timeouts when installing.

**Reply by B4nan** [reply]
Or maybe its something with env vars, I tried your repro, and when I comment out the import for your config file (which only throws since there is no env file), and tried uinsg the CLI, it works fine on my end, I can see it finds the TS config and reads it.

```
➜  classifieds-service-backend-typescript git:(main) ✗ npx mikro-orm-esm debug
Current MikroORM CLI configuration
 - dependencies:
   - mikro-orm 6.3.0
   - node 22.4.1
   - knex 3.1.0
   - pg 8.12.0
   - typescript 5.5.3
 - package.json found
 - ts-node enabled
 - searched config paths:
   - /Users/adamek/htdocs/mikro-orm/classifieds-service-backend-typescript/src/mikro-orm.config.ts (found)
   - /Users/adamek/htdocs/mikro-orm/classifieds-service-backend-typescript/mikro-orm.config.ts (not found)
   - /Users/adamek/htdocs/mikro-orm/classifieds-service-backend-typescript/src/mikro-orm.config.js (not found)
   - /Users/adamek/htdocs/mikro-orm/classifieds-service-backend-typescript/mikro-orm.config.js (not found)
 - configuration found
 - database connection failed (database "test" does not exist)
 - will use `entities` array (contains 0 references and 1 paths)
   - /Users/adamek/htdocs/mikro-orm/classifieds-service-backend-typescript/dist/**/*.entity.js (not found)
 - could use `entitiesTs` array (contains 0 references and 1 paths)
   - /Users/adamek/htdocs/mikro-orm/classifieds-service-backend-typescript/src/**/*.entity.ts (found)
```

**Reply by sz3lbi** [reply]
Okay, I also used the CLI and this is really interesting...
```
pnpm mikro-orm-esm debug
Current MikroORM CLI configuration
 - dependencies:
   - mikro-orm 6.3.0
   - node 22.5.1
   - knex 3.1.0
   - pg 8.12.0
   - typescript 5.5.3
 - package.json found
 - ts-node enabled
 - searched config paths:
   - /home/username/repos/classifieds-service-backend-typescript/src/mikro-orm.config.ts (found)
   - /home/username/repos/classifieds-service-backend-typescript/mikro-orm.config.ts (not found)
   - /home/username/repos/classifieds-service-backend-typescript/src/mikro-orm.config.js (not found)
   - /home/username/repos/classifieds-service-backend-typescript/mikro-orm.config.js (not found)
 - configuration found
 - database connection successful
 - will use `entities` array (contains 0 references and 1 paths)
   - /home/username/repos/classifieds-service-backend-typescript/dist/**/*.entity.js (not found)
 - could use `entitiesTs` array (contains 0 references and 1 paths)
   - /home/username/repos/classifieds-service-backend-typescript/src/**/*.entity.ts (found)
```
But when I run it with the `package.json` script:
```
pnpm run dev

> classifieds-service-backend@0.2.0 dev /home/username/repos/classifieds-service-backend-typescript
> nodemon src/index.ts

[nodemon] 3.1.4
[nodemon] to restart at any time, enter `rs`
[nodemon] watching path(s): src/**/*
[nodemon] watching extensions: ts
[nodemon] starting `node --no-warnings=ExperimentalWarning --loader ts-node/esm src/index.ts`
/home/username/repos/classifieds-service-backend-typescript/node_modules/.pnpm/@mikro-orm+core@6.3.0/node_modules/@mikro-orm/core/utils/ConfigurationLoader.js:44
        throw new Error(`MikroORM config file not found in ['${paths.join(`', '`)}']`);
              ^

Error: MikroORM config file not found in ['./src/mikro-orm.config.js', './mikro-orm.config.js']
    at ConfigurationLoader.getConfiguration (/home/username/repos/classifieds-service-backend-typescript/node_modules/.pnpm/@mikro-orm+core@6.3.0/node_modules/@mikro-orm/core/utils/ConfigurationLoader.js:44:15)
    at PostgreSqlMikroORM.init (/home/username/repos/classifieds-service-backend-typescript/node_modules/.pnpm/@mikro-orm+core@6.3.0/node_modules/@mikro-orm/core/MikroORM.js:32:58)

Node.js v22.5.1
[nodemon] app crashed - waiting for file changes before starting...
```
Direct execution:
```
pnpm node --no-warnings=ExperimentalWarning --loader ts-node/esm src/index.ts
/home/username/repos/classifieds-service-backend-typescript/node_modules/.pnpm/@mikro-orm+core@6.3.0/node_modules/@mikro-orm/core/utils/ConfigurationLoader.js:44
        throw new Error(`MikroORM config file not found in ['${paths.join(`', '`)}']`);
              ^

Error: MikroORM config file not found in ['./src/mikro-orm.config.js', './mikro-orm.config.js']
    at ConfigurationLoader.getConfiguration (/home/username/repos/classifieds-service-backend-typescript/node_modules/.pnpm/@mikro-orm+core@6.3.0/node_modules/@mikro-orm/core/utils/ConfigurationLoader.js:44:15)
    at PostgreSqlMikroORM.init (/home/username/repos/classifieds-service-backend-typescript/node_modules/.pnpm/@mikro-orm+core@6.3.0/node_modules/@mikro-orm/core/MikroORM.js:32:58)

Node.js v22.5.1
```
In the meantime, I've updated Node to `22.5.1`.

**Reply by sz3lbi** [reply]
I've tried using `npm` instead of `pnpm`, but it gives exactly the same output.
```
npx mikro-orm-esm debug
Current MikroORM CLI configuration
 - dependencies:
   - mikro-orm 6.3.0
   - node 22.5.1
   - knex 3.1.0
   - pg 8.12.0
   - typescript 5.5.4
 - package.json found
 - ts-node enabled
 - searched config paths:
   - /home/username/repos/classifieds-service-backend-typescript/src/mikro-orm.config.ts (found)
   - /home/username/repos/classifieds-service-backend-typescript/mikro-orm.config.ts (not found)
   - /home/username/repos/classifieds-service-backend-typescript/src/mikro-orm.config.js (not found)
   - /home/username/repos/classifieds-service-backend-typescript/mikro-orm.config.js (not found)
 - configuration found
 - database connection successful
 - will use `entities` array (contains 0 references and 1 paths)
   - /home/username/repos/classifieds-service-backend-typescript/dist/**/*.entity.js (not found)
 - could use `entitiesTs` array (contains 0 references and 1 paths)
   - /home/username/repos/classifieds-service-backend-typescript/src/**/*.entity.ts (found)
```
```
npm run dev

> classifieds-service-backend@0.2.0 dev
> nodemon src/index.ts

[nodemon] 3.1.4
[nodemon] to restart at any time, enter `rs`
[nodemon] watching path(s): src/**/*
[nodemon] watching extensions: ts
[nodemon] starting `node --no-warnings=ExperimentalWarning --loader ts-node/esm src/index.ts`
/home/username/repos/classifieds-service-backend-typescript/node_modules/@mikro-orm/core/utils/ConfigurationLoader.js:44
        throw new Error(`MikroORM config file not found in ['${paths.join(`', '`)}']`);
              ^

Error: MikroORM config file not found in ['./src/mikro-orm.config.js', './mikro-orm.config.js']
    at ConfigurationLoader.getConfiguration (/home/username/repos/classifieds-service-backend-typescript/node_modules/@mikro-orm/core/utils/ConfigurationLoader.js:44:15)
    at PostgreSqlMikroORM.init (/home/username/repos/classifieds-service-backend-typescript/node_modules/@mikro-orm/core/MikroORM.js:32:58)

Node.js v22.5.1
[nodemon] app crashed - waiting for file changes before starting...
```

**Reply by sz3lbi** [reply]
As for the `.env` file, it needs two properties: `DATABASE_URL` which is a PostgreSQL connection URL and `SECRET_KEY` which is a string with a minimum length of 50 characters. To generate my secret I use the command `openssl rand -hex 32`.

**Reply by B4nan** [reply]
Now I see what you do differently:

https://github.com/mikro-orm/guide/blob/master/src/db.ts#L28

There is an explicit import of the config file, while in your repo you skipped that part.

**Reply by sz3lbi** [reply]
OK, after importing it explicitly and passing it in as a parameter, the application launches without any problems.  
However, [according to the docs](https://mikro-orm.io/docs/guide/first-entity#initializing-the-orm): "_We used the init() method without any parameters, which results in the ORM loading the CLI config automatically_". I guess this should be the case here too? Especially considering that this was the case in `6.2.9` and the CLI sees the file?

**Reply by B4nan** [reply]
This should help eb100fec7b06deb1e653b594f730c901ec8ce8e7

**Reply by Ashvith10** [reply]
From what I've noticed, without building the project at least once, after updating the entity schema or after clearing the build files, it won't work. Is this a weird behavior?

**Reply by B4nan** [reply]
That only means you got things misconfigured, TS files are always preferred, as long as you have `ts-node` installed.

edit: I mean for CLI, if you are talking about running your app, it's up to you to register TS support, the ORM won't do that.

**Reply by Ashvith10** [reply]
@B4nan It looks like I have the same problem that @sz3lbi had, even after using the new updated version. Here's the debug log:

```console
$ npx mikro-orm-esm debug
Current MikroORM CLI configuration
 - dependencies:
   - mikro-orm 6.4.1
   - node 18.19.0
   - typescript 5.7.2
 - package.json found
 - ts-node enabled
 - searched config paths:
   - /home/ashvith/Desktop/blog-api/src/mikro-orm.config.ts (found)
   - /home/ashvith/Desktop/blog-api/mikro-orm.config.ts (not found)
   - /home/ashvith/Desktop/blog-api/src/mikro-orm.config.js (not found)
   - /home/ashvith/Desktop/blog-api/mikro-orm.config.js (not found)
 - searched for config name: default
 - configuration found
 - driver dependencies:
   - knex 3.1.0
   - sqlite3 5.1.7
 - database connection successful
 - `preferTs` flag explicitly set to true, will use `entitiesTs` array (this value should be set to `false` when running compiled code!)
 - could use `entities` array (contains 0 references and 1 paths)
   - /home/ashvith/Desktop/blog-api/dist/**/*.entity.js (not found)
 - will use `entitiesTs` array (contains 0 references and 1 paths)
   - /home/ashvith/Desktop/blog-api/src/**/*.entity.ts (found)
```

Only after running build once does the issue get resolved for me, if I don't implicitly set the config, as you've mentioned in https://github.com/mikro-orm/guide/blob/master/src/db.ts#L28.

**Reply by B4nan** [reply]
So you get this with the https://github.com/mikro-orm/guide repo? It works just fine for me, you will need to share more details on what you are doing. That project is not using EntitySchema, so you must have done some adjustments.

**Reply by B4nan** [reply]
Note that you are not supposed to be using ts-morph with EntitySchema, there is no point in doing that.

**Reply by Ashvith10** [reply]
> So you get this with the https://github.com/mikro-orm/guide repo? It works just fine for me, you will need to share more details on what you are doing. That project is not using EntitySchema, so you must have done some adjustments.

I was using EntitySchema for some other project, but then I decide to complete the first two chapters of the guide for some basic understanding of the ORM, so apologies for the confusion.

Weirdly enough, after (I think I may be wrong) deleting the `dist` file and the `temp` file (I didn't delete temp before), it worked just fine? I am still not sure what the actual issue is, and I am not able to reproduce this.
